File: 400/414 - Third Maximum Number.c
#include <limits.h>

int thirdMax(int* nums, int numsSize) {
    long long first = LLONG_MIN, second = LLONG_MIN, third = LLONG_MIN;
    int i, num;
    
    for (i = 0; i < numsSize; ++i) {
        num = nums[i];
        if (num > first) {
            third = second;
            second = first;
            first = num;
        } else if (num > second && num < first) {
            third = second;
            second = num;
        } else if (num > third && num < second)
            third = num;
    }
    
    return (third == LLONG_MIN || third == second) ? (int)first : (int)third; 
}

File: csharp/414.third-maximum-number.cs
/*
 * @lc app=leetcode id=414 lang=csharp
 *
 * [414] Third Maximum Number
 */

using System.Collections.Generic;

public class Solution {
    public int ThirdMax(int[] nums) {
        IList<int> max = new List<int>(3);
        foreach (int num in nums) {
            int i;
            bool flag = false;
            for (i = 0; i < max.Count; ++i) {
                if (max[i] < num) {
                    break;
                } else if (max[i] == num) {
                    flag = true;
                    break;
                }
            }
            if (flag) {
                continue;
            }
            if (max.Count < 3) {
                max.Insert(i, num);
            } else {
                if (i < max.Count) {
                    max.RemoveAt(2);
                    max.Insert(i, num);
                }
            }
        }
        return max[max.Count == 3 ? 2 : 0];
    }
}

